templates/setup/config/local.js

```diff
@@ -37,14 +37,38 @@ module.exports = {
    },
    /* end datastores */
 
+   /** Session */
+   session: {
+      secret: "<%=secret%>",
+      // Session secret is automatically generated during install.
+      // Replace at your own risk in production-- you will invalidate the cookies
+      // of your users, forcing them to log in again.     
+   },
+   <% if (!develop) { %>
+   socket: {
+      /**
+       * @function beforeConnect
+       * This allows all socket connections. It is recommended to customize
+       * this in production or use `onlyAllowOrigins` instead.
+       * Sails won't lift in production mode without of if these set.
+       * see https://sailsjs.com/documentation/reference/configuration/sails-config-sockets#sails-config-sockets
+       */
+      beforeConnect: function (handshake, proceed) {
+         console.warn(
+            "Allowing all socket connections. This should be changed."
+         );
+         return proceed(undefined, true);
+      },
```

Comment:
So are we planning to develop this further in the future? `onlyAllowOrigins` and `beforeConnect` seem like newish Sails features that we haven't encountered on our previous sites. If I understand correctly, what we're doing here is just bypassing the new Sails security measure.

Presumably, the Sails devs felt that socket connections originating from 3rd party sites/domains are a security risk, and were closing the hole by whitelisting domains through the `onlyAllowOrigins` setting. I wasn't aware of this previously, but maybe [something like this](https://www.google.com/search?q=cross+site+socket+connection+vulnerability&oq=cross+site+socket+connection+vulnerability&aqs=chrome..69i57j69i64.16669j0j4&sourceid=chrome&ie=UTF-8).

And of course, adding a domain whitelist in our case is tricky because the list of tenants can change as time goes by. I'm just promoting awareness now. I don't think any of our sites are in urgent danger from this type of attack.

Reply:
Counterpoint to my previous comment: none of our Sails sites have ever been protected against cross-site websocket attacks. So bypassing the new Sails security measure doesn't make us any worse off than before.

Reply:
is there a drawback to actually implementing this security feature on our site?

Reply:
It's probably quite minor, but I think the drawback is some added complexity to our workflow, or some added wait time per websocket request.

From what I gather, enabling this feature for normal sites is as simple as adding this to the config:
```js
socket: {
    onlyAllowOrigins: "mysite.example.com,othersite.example.com"
}
```

But each AppBuilder instance can support many different domains, and more can be added over time. It's even possible to add new tenants without restarting AppBuilder, though that's probably not a critical feature to have. So we could manually change the `onlyAllowOrigins` config setting whenever new domains are added, and then restart.

Another possibility is to automate this by referencing the site URLs in the `appbuilder-admin.site_tenant` table. Maybe something like:

```js
socket: {
    beforeConnect: function (handshake, proceed) {
        // Build set of tenant URLs.
        // (Look up DB every time?)
        const tenantURLs = new Set();
        const results = await db.query(`SELECT something FROM site_tenant`);
        for (let row of results) {
            tenantURLs.add(row.url);
        }
        
        // Get request origin from handshake?
        const origin = handshake.someMethod.orProperty;

        if (tenantURLs.has(origin)) {
            return proceed(undefined, true);
        } else {
            return proceed(new Error("Nope"));
        }
    }
}
```

I guess it could be fleshed out properly as its own sprint task in the future.

Reply:
My thought, initially, was to leave this up to the instance to set it's allowed origins in `local.js`, since each will look different. I think it should be possible to add a `beforeConnect` function to support `*.domain.com` so we can add new tenants without changing the config.

I like the idea of using the tenantURLs, but not sure about making a DB request before each connection.